lib/remap-functions/src/length.rs

```diff
@@ -0,0 +1,144 @@
+use remap::prelude::*;
+use std::convert::TryFrom;
+
+#[derive(Clone, Copy, Debug)]
+pub struct Length;
+
+impl Function for Length {
+    fn identifier(&self) -> &'static str {
+        "length"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            accepts: |v| matches!(v, Value::Array(_) | Value::Map(_)),
+            required: true,
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Result<Box<dyn Expression>> {
+        let value = arguments.required("value")?.boxed();
+
+        Ok(Box::new(LengthFn { value }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct LengthFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for LengthFn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        use Value::*;
+
+        let value = self.value.execute(state, object)?;
+
+        match value {
+            Array(v) => safe_i64(v.len()),
+            Map(v) => safe_i64(v.len()),
+            _ => Err("unsupported type".into()),
+        }
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        use value::Kind;
+
+        self.value
+            .type_def(state)
+            .fallible_unless(Kind::Array | Kind::Map)
+            .with_constraint(Kind::Integer)
+    }
+}
+
+fn safe_i64(n: usize) -> Result<Value> {
+    match i64::try_from(n) {
+        Ok(n) => Ok(Value::from(n)),
+        Err(_) => Err("64-bit integer overflow".into()),
+    }
+}
```

Comment:
If we do this, then technically this function is always fallible. I propose we just let it wrap around the integer if this happens, since we're optimizing here for the observability use-case, and I've literally never seen anyone send an array, map or string with a size greater than 9223372036854775807 in their events 😄

I know, I know "but that's incorrect behavior!", we can document it (this function wraps the integer if it's larger than ...) and I guarantee you (famous last words) that we'll never hear anyone complain.

It's just too annoying to have to constantly handle error cases for such an extreme edge-case that I think it's the only correct thing to do here.

Reply:
:+1: I wasn't sure what would happen if we return an error here even if the function is infallible for the given type. Would it make sense to just `panic` here? I realize the case is extreme so users are unlikely to hit it either way.

Reply:
I would prefer not to introduce any panicking code if at all possible. I think documenting it makes it so that (a) people can read about it and (b) we can link to it if anyone _ever_ hits this case.

Reply:
:+1: I was just thinking that the documentation around this error is likely to be mostly noise given the extremeness of this case. I'm fine with this as-is though.

Reply:
I would actually document it in the "literals" section of the website for integers and floats. Basically, stating that we use signed 64-bit integers (and floats) and that those integers are wrapped if they exceed their maximum value.